src/VectorBundle.jl

```diff
@@ -0,0 +1,372 @@
+
+"""
+    VectorSpaceType
+
+Abstract type for tangent space, cotangent space, their tensor products,
+exterior products etc.
+
+Every vector space `VS` is supposed to provide:
+* a method of constucting vectors,
+* basic operations: addition, subtraction, multiplication by a scalar
+  and negation (unary minus),
+* zero_vector!(VS, v, x) to construct zero vectors at point `x`,
+* `similar(v, T)` for vector `v`,
+* `copyto!(v, w)` for vectors `v` and `w`,
+* `eltype(v)` for vector `v`,
+* `vector_space_dimension(::VectorBundleFibers{<:typeof(VS)})`.
+
+Optionally:
+* inner product via `inner` (used to provide Riemannian metric on vector
+  bundles),
+* flat_isomorphism! and sharp_isomorphism!,
+* norm (by default uses `inner`),
+* project_vector! (for embedded vector spaces),
+* representation_size (if support for `ProductArray` is desired),
+* broadcasting for basic operations.
+"""
+abstract type VectorSpaceType end
+
+struct TangentSpaceType <: VectorSpaceType end
+struct CotangentSpaceType <: VectorSpaceType end
+
+TCoTSpaceType = Union{TangentSpaceType, CotangentSpaceType}
+
+const TangentSpace = TangentSpaceType()
+const CotangentSpace = CotangentSpaceType()
+
+"""
+    TensorProductType(spaces::VectorSpaceType...)
+
+Vector space type corresponding to the tensor product of given vector space
+types.
+"""
+struct TensorProductType{TS<:Tuple} <: VectorSpaceType
+    spaces::TS
+end
+
+TensorProductType(spaces::VectorSpaceType...) = TensorProductType{typeof(spaces)}(spaces)
+
+"""
+    VectorBundleFibers(VS::VectorSpaceType, M::Manifold)
+
+Type representing a family of vector spaces (fibers) of a vector bundle over `M`
+with vector spaces of type `VS`. In contrast with `VectorBundle`, operations
+on `VectorBundleFibers` expect point-like and vector-like parts to be
+passed separately instead of being bundled together. It can be thought of
+as a representation of vector spaces from a vector bundle but without
+storing the point at which a vector space is attached (which is specified
+separately in various functions).
+"""
+struct VectorBundleFibers{TVS<:VectorSpaceType, TM<:Manifold}
+    VS::TVS
+    M::TM
+end
+
+"""
+    VectorSpaceAtPoint(fiber::VectorBundleFibers, x)
+
+A vector space (fiber type `fiber` of a vector bundle) at point `x` from
+the manifold `fiber.M`.
+"""
+struct VectorSpaceAtPoint{TFiber<:VectorBundleFibers, TX}
```

Comment:
Would be nice to have convenient types and constructors for `TangentSpaceAtPoint` and `CotangentSpaceAtPoint`, similar to `TangentBundleFibers` defined below.